### PR TITLE
[SYCLomatic] Execute device synchronization before device pointer free

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -10580,9 +10580,16 @@ void MemoryMigrationRule::freeMigration(const MatchFinder::MatchResult &Result,
       std::ostringstream Repl;
       buildTempVariableMap(Index, C, HelperFuncType::HFT_DefaultQueue);
       if (hasManagedAttr(0)(C)) {
-          ArgStr = "*(" + ArgStr + ".get_ptr())";
+        ArgStr = "*(" + ArgStr + ".get_ptr())";
       }
-      Repl << MapNames::getClNamespace() + "free(" << ArgStr
+      std::string NewFuncName;
+      if (DpctGlobalInfo::useNoQueueDevice()) {
+        NewFuncName = MapNames::getClNamespace() + "free";
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: may need a `{{NEEDREPLACEQ}}.wait_and_throw()` before the `sycl::free`?

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -10580,9 +10580,16 @@ void MemoryMigrationRule::freeMigration(const MatchFinder::MatchResult &Result,
...
+      } else {
+        requestFeature(HelperFeatureEnum::device_ext);
+        NewFuncName = MapNames::getDpctNamespace() + "dpct_free";
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: can use `Repl <<` directly, needn't the `NewFuncName` variable.

File: clang/test/dpct/vector_add_format.cu [hunk 3]
```diff
@@ -54,13 +54,6 @@ int main()
   float Result[VECTOR_SIZE] = {};
   cudaMemcpy(Result, d_C, VECTOR_SIZE * sizeof(float), cudaMemcpyDeviceToHost);
 
-  //     CHECK:  sycl::free(d_A, q_ct1);
-  // CHECK-NEXT:  sycl::free(d_B, q_ct1);
-  // CHECK-NEXT:  sycl::free(d_C, q_ct1);
-  cudaFree(d_A);
```
Comment (tangjj11) [inline clang/test/dpct/vector_add_format.cu:60]: can change the `CHECK` but not delete.

**Review by tangjj11 — CHANGES_REQUESTED** [review]

**Review by tangjj11 — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/memory.hpp [hunk 4]
```diff
@@ -832,6 +832,7 @@ static inline void *dpct_malloc(size_t &pitch, size_t x, size_t y,
 /// \returns no return value.
 static inline void dpct_free(void *ptr,
                              sycl::queue &q = get_default_queue()) {
+  dpct::get_current_device().queues_wait_and_throw();
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/memory.hpp:836]: L835 is not required in USM_none

File: clang/test/dpct/cub/devicelevel/device_reduce_arg_max.cu [hunk 5]
```diff
@@ -34,7 +34,7 @@ void test1() {
 // CHECK-NOT: size_t temp_storage_bytes = 0;
 // CHECK: DPCT1026:{{.*}}: The call to cub::DeviceReduce::ArgMax was removed because this functionality is redundant in SYCL.
 // CHECK: dpct::reduce_argmax(oneapi::dpl::execution::device_policy(q_ct1), d_in, d_out, num_items);
-// CHECK-NOT: sycl::free({{.*}})
+// CHECK-NOT: dpct::dpct_free({{.*}})
```
Comment (zhimingwang36) [inline clang/test/dpct/cub/devicelevel/device_reduce_arg_max.cu:37]: Please config migrate to sycl::free  if "---optimize-migration" is specified.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**